topic_sidebar: Add unmute topic option in development environment.

Ddharmani3 opened the pull request:
if in development environment, for a topic in muted stream show additional option for unmuting the topic that sets visibility policy to unmute.

<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->
#24244 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
setting and removing unmute visibility policy:

https://user-images.githubusercontent.com/64723994/229842881-4ae9ec42-0404-4988-8eef-8419addae6d2.mp4

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/muted_topics_ui.js:
```diff
@@ -113,6 +113,19 @@ export function toggle_topic_mute(message) {
+    channel.post({
+        url: "/json/user_topics",
+        data,
+    });
```
You'll very likely want success/error handlers here.

timabbott commented on web/src/stream_popover.js:
```diff
@@ -604,6 +610,22 @@ export function register_stream_handlers() {
+
+        const topic = $(e.currentTarget).attr("data-topic-name");
+        const new_visibility_policy =
+            user_topics.get_topic_visibility_policy(stream_id, topic) === 2 ? 0 : 2;
```
We shouldn't ever use raw numbers that translate to logical states like the 0/2 in code like this -- it's impossible to read. See `web/src/settings_config.ts` for some examples of a good pattern; I see we actually already define the right structure in `user_topics.js`:

```
export const visibility_policy = {                                                                      
    INHERIT: 0,                                                                                         
    MUTED: 1,                                                                                           
    UNMUTED: 2,                                                                                         
    FOLLOWED: 3,                                                                                        
};                                                                                                      
```

timabbott commented on web/src/user_topics.js:
```diff
@@ -73,9 +100,30 @@ export function set_user_topic(user_topic) {
+                topic,
+                date_muted,
+                user_topic.visibility_policy,
+            );
```
You should be deleting the `add_muted_topic` calls here -- muted topics is a special case of visibility policies, so you're doing the same thing twice in a row with different API endpoints here.

timabbott commented on web/templates/topic_sidebar_actions.hbs:
```diff
@@ -8,6 +8,22 @@
+                <a tabindex="0" class="sidebar-popover-toggle-topic-visibility" data-stream-id="{{ stream_id }}" data-topic-name="{{ topic_name }}">
+                    <i class="zulip-icon zulip-icon-mute" aria-hidden="true"></i>
+                    {{#if topic_unmuted}}
+                        {{t "Specially unmute topic"}}
```
This can just be "Unmute topic".

timabbott commented on web/src/stream_popover.js:
```diff
@@ -245,9 +248,12 @@ function build_topic_popover(opts) {
         topic_muted,
+        topic_unmuted,
         can_move_topic,
+        dev_env: page_params.development_environment,
```
Call this `development_environment`; we use that name in many templates and it makes grepping way more convenient.

(In general, one should try to avoid renaming a property to a slightly different name like this when passing it between different parts of the system -- it's almost always bad for readability and has no real benefit).

timabbott commented on web/src/stream_popover.js:
```diff
@@ -604,6 +610,22 @@ export function register_stream_handlers() {
 }
 
 export function register_topic_handlers() {
+    // Toggle topic visibility_policy (currently in dev environment)
```
Remove the part about "currently in dev ..." in the comment here. Writing comments like that is a bad idea because they almost always end up stale, since it's not next to the code that only enables this feature in the development environment, so when we do enable it for everyone, we'll very likely fail to update it here.

timabbott commented on web/src/user_topics.js:
```diff
@@ -14,6 +15,32 @@ export const visibility_policy = {
+    }
+    const sub_dict = topic_visibility_policies.get(stream_id);
+    if (sub_dict && sub_dict.get(topic)) {
+        return sub_dict.get(topic)[1];
```
What does this `[1]` do? We very likely want to use an object here.

Ddharmani3 commented:
@timabbott, Pushing the progress till now, I have addressed all the feedback received above except the name of "Unmute topic". This PR is not ready yet as it has a few failing tests that I will fix and push the updated PR tomorrow till then you can review the PR again for any additional feedback if you like. I'll also address additional feedback alongside the failing tests.

prakhar1144 commented on web/src/settings_config.ts:
```diff
@@ -366,6 +366,13 @@ export const waiting_period_threshold_dropdown_values = [
 
 export const retain_message_forever = -1;
 
+export const topic_visibility_policy = {
```
We already have
```
export const visibility_policy = {
    INHERIT: 0,
    MUTED: 1,
    UNMUTED: 2,
    FOLLOWED: 3,
};
``` 
in `web/src/user_topics.js`.

Any particular reason for including it here?

Ddharmani3 replied:
As other similar objects that store the API values were in settings_config.ts. so I think this should also be here.

timabbott replied:
We don't need a duplicate, `user_topics` is the module that owns this data, so it belongs there, so we can delete this copy.

(`settings_config` is for data that is settings).

prakhar1144 commented on web/src/muted_topics_ui.js:
```diff
@@ -44,18 +45,28 @@ export function handle_topic_updates(user_topic) {
+        url: "/json/user_topics",
         data,
         success() {
+            // Accidentally unmuting a topic isn't as much an issue as accidentally muting
```
This comment should be updated, considering the fact that we have now other visibility policies.

Ddharmani3 replied:
I will update this comment. Any suggestions for how to handle unmuted topic success?

Ddharmani3 replied:
I think unmute topic success can be handled when working on hotkeys as the feedback_widget is only shown for the mute topic if it is done from hotkey.

prakhar1144 commented on web/src/muted_topics_ui.js:
```diff
@@ -84,41 +100,40 @@ export function mute_topic(stream_id, topic, from_hotkey) {
+        update_topic_visibility_policy(
+            stream_id,
+            topic,
+            settings_config.topic_visibility_policy.UNMUTED,
```
it should set visibility_policy to `INHERIT`, not `UNMUTED`

prakhar1144 commented on web/src/muted_topics_ui.js:
```diff
@@ -84,41 +100,40 @@ export function mute_topic(stream_id, topic, from_hotkey) {
-    });
-}
-
 export function toggle_topic_mute(message) {
```
In follow-up PR, we should maybe generalize this to, 
`toggle_visibility_policy` or `add_or_remove_visibility_policy`. Namings we can discuss.

(I think, this handles hotkey-toggling, so this should be renamed when working on hotkeys for UNMUTE.)

Ddharmani3 replied:
Yes, this can be done when working on hotkeys for UNMUTE as a followup to this PR.

prakhar1144 commented on web/src/muted_topics_ui.js:
```diff
@@ -84,41 +100,40 @@ export function mute_topic(stream_id, topic, from_hotkey) {
     }
 }
 
 export function mute_or_unmute_topic($elt, mute) {
```
This function should be generalized to handle other visibility policies, in a follow-up PR.

prakhar1144 reviewed (commented):
@Ddharmani3  Thanks for working on it!

Reviewed first 2 commits.

Ddharmani3 commented on web/src/muted_topics_ui.js:
```diff
@@ -44,18 +45,28 @@ export function handle_topic_updates(user_topic) {
 
-export function mute_topic(stream_id, topic, from_hotkey) {
-    const stream_name = stream_data.maybe_get_stream_name(stream_id);
+export function update_topic_visibility_policy(
```
I wanted to do it too but for that, a new frontend test should be added and there was no test for mute/unmute topic functions in `muted_topics_ui.js`. There is no test file for `muted_topics_ui.js` and it is in EXEMPT_FILES in test-js-with-node. should I add `web/src/user_topics.js` in EXEMPT_FILES too and shift the `update_topic_visibility_policy` function to `user_topics.js`. We can write a test for this function later along with new tests that will be written for #24243 as this PR need to be merged soon in order to start work for #24243.

@prakhar1144 @timabbott your thoughts?

timabbott replied:
> should I add web/src/user_topics.js in EXEMPT_FILES too and shift the update_topic_visibility_policy function to user_topics.js

Yes, we should always prioritize the actual code working how we want it to over testing alerts like `EXEMPT_FILES`, which exists to help us notice when we lose test coverage on a module.

Ddharmani3 commented:
I have updated all the instances where I mistakenly wrote UNMUTED instead of INHERIT due to some naming confusion. 
I am waiting for feedback on [this](https://github.com/zulip/zulip/pull/24989#discussion_r1158660954).

I think there are only 2 failing tests: one mentioned above that will arise due to lost coverage if we move the `update_topic_visibility_policy` function to `user_topics.js`  and the second in `settings_muted_topics.test.js` as mentioned in this [commit](74335594b7d93329948e86195d0730946ae7222f).

Both of them can be addressed while working on #24243 as this PR should get merged to start working on 6 parallel UI tasks of #24243.

@timabbott your feedback on this?

Ddharmani3 commented:
Summary:

* Progress till now:
       * commit 1[b475046ec241481323993ce1990a155dcbde2bbd]: Added `topic_visibility_policy` object containing {visibility_policy: corresponding integer value} pairs in `settings_config.ts`.
       * commit 2[4b4ab7478541073c7b277afb10d68394feca8498]: Added new function `update_topic_visibility_policy` that uses new API endpoint. Removed `mute_topics` and `unmute_topic` functions that were using the old API endpoint.
       * commit 3[52b2682aed3a3a3e729c4264388ef6eff4bbeb38]: * Replaced `muted_topics` Map with `topics_dict` Map. 
                           * Removed `add_muted_topic` and `remove_muted_topic` functions and replaced them with a single function 
                              `update_topics_dict` that add/remove/update topics in new `topics_dict` Map. 
                           * Added get_topic_visibility_policy which returns visibility_policy if stream_id is not undefined else return false.
       * commit 4[3f22eed8e226482222d5e05316d73547d296a6a6]: If in development environment, in topic_sidebar_action popover, include an option to add/remove unmute visibility_policy for a topic in a muted stream. Updated click handlers for both mute/remove mute and unmute/remove unmute options.
       * commit 5[9b23ac0eaaec461ff77eaf7d4bdb107ae11aacc6]: Updated the following node tests to use updated functions.
      
* Discussion required:
       * The naming discussion for both the [design part](https://chat.zulip.org/#narrow/stream/101-design/topic/unmuting.20in.20muted.20streams/near/1541291) and the [frontend technical](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Unmute.20Topic.20UI.20.2324244.20.26.20.2324243/near/1541753) part is CZO.
       * In commit 2, the new function `update_topic_visibility_policy` which uses new API endpoint makes `user_topics.js` lose test 
          coverage. The functions `mute_topic` and `unmute_topic` in `muted_topics_ui.js` that were using the old API endpoint had 
          no tests and `muted_topics_ui.js` was in EXEMPT_FILES in `test-js-with-node`. Was this intentional? Should I also keep 
          `user_topics.js` in EXEMPT_FILES? I have kept user_topics.js in EXEMPT_FILES in this PR till further feedback.
       * Review whether the test in `settings_muted_topic.test.js` is updated correctly?

* Work to be done in follow-ups:
        * In commit 3, I have created 2 helper functions is_topic_muted and is_topic_unmuted for this PR only that use 
          `get_topic_visibility_policy` to check for mute and unmute respectively. `is_topic_muted` is kept as it has 40+ occurrences 
          in the repo. But in the follow-up, these functions will be converted to one general function something like 
          `has_visibility_policy` that returns true if a topic has visibility_policy passed as a parameter.
        * Make functions more generalized to work for all visibility policies rather than functions like `mute_or_unmute_topic` and 
          `mute_or_unmute_topic` in `muted_topics_ui.js`.
        * Work on UI tasks of #24243 once this PR is merged.

prakhar1144 commented on web/src/user_topics.js:
There are some typos in the last paragraph of the commit message.
```
**Relaced** switch case statement which was using add_muted_topic
**an** remove_muted_topic functions in set_user_topic with new function
update_topics_dict.
```

Ddharmani3 replied:
I will correct typo in the next push

prakhar1144 commented on web/src/user_topics.js:
```diff
@@ -11,54 +11,67 @@ import * as stream_data from "./stream_data";
 import {get_time_from_date_muted} from "./util";
 
-const muted_topics = new Map();
+const topics_dict = new Map();
```
We can maybe use `user_topics` instead of `topics_dict` -- because it doesn't store data about every topic.

Ddharmani3 replied:
I will wait for feedback on this name in [CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Unmute.20Topic.20UI.20.2324244.20.26.20.2324243/near/1542446) before changing.

prakhar1144 commented on web/src/user_topics.js:
```diff
@@ -11,54 +11,67 @@ import * as stream_data from "./stream_data";
-export function remove_muted_topic(stream_id, topic) {
-    const sub_dict = muted_topics.get(stream_id);
-    if (sub_dict) {
+    const time = get_time_from_date_muted(date_updated);
```
I think we should rename `get_time_from_date_muted` in a prep commit, to update the `date_muted` part of the name.

Ddharmani3 replied:
Sure, I will do that once the name gets finalized in the design stream :)

prakhar1144 commented on web/src/user_topics.js:
```diff
@@ -11,54 +11,67 @@ import * as stream_data from "./stream_data";
-    let sub_dict = muted_topics.get(stream_id);
+export function update_topics_dict(stream_id, topic, visibility_policy, date_updated) {
+    let sub_dict = topics_dict.get(stream_id);
     if (!sub_dict) {
```
This will have incorrect behavior when `!sub_dict is true` and `visibility_policy=INHERIT`

Ddharmani3 replied:
Refactored this function to have the correct behavior as before in the updated PR.

prakhar1144 reviewed (commented):
For the last commit,

As mentioned in the [docs](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-discipline):

> It should pass tests (so test updates needed by a change should be in the same commit as the original change, 

So, tests should be included in the respective commits, where changes occured.

Ddharmani3 commented:
I have updated the PR to resolve conflicts that were raised due to the topic menu being converted to Tippy. Updated PR to work for the tippy menu too.
Also updated above [summary](https://github.com/zulip/zulip/pull/24989#issuecomment-1498011108) as per updated PR.
I will update the namings once finalized on CZO.

timabbott commented on web/src/user_topics.js:
```diff
@@ -58,6 +64,62 @@ export function get_muted_topics() {
+            if (new_visibility_policy === settings_config.topic_visibility_policy.INHERIT) {
+                feedback_widget.dismiss();
+                return;
+            }
```
I don't understand why we'd need to dismiss the feedback widget here -- is this if one toggles it back while it's still visible?

Ddharmani3 replied:
The feedback_widget.dismiss() function was originally included in the unmute_topic function. However, upon checking it today, I realized that feedback_widget.dismiss() does the same thing as clicking on "undo mute" since it calls animate.fadeOut(). 
Therefore, I removed the if condition in the updated PR.

prakhar1144 replied:
> is this if one toggles it back while it's still visible?

**Yes**

@Ddharmani3 after your recent push, the widget is not dismissed on toggling back (while it's visible) (using hotkey).

timabbott replied:
Yeah I think my comment was in error, and we actually do need this line.

Ddharmani3 commented:
I have updated the PR to reflect the suggested changes in naming, as discussed [here](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Unmute.20Topic.20UI.20.2324244.20.26.20.2324243/near/1543398).
Additionally, I have removed the `topic_visibility_policy` from `settings_config.ts` and used the original `visibility_policy` from `user_topics.js`, as was suggested [here](https://github.com/zulip/zulip/pull/24989#discussion_r1160999351).

This PR is ready for review now.

prakhar1144 commented on web/src/user_topics.js:
```diff
@@ -19,45 +19,58 @@ export const visibility_policy = {
+    return visibility_policy.INHERIT;
+}
+
+export function is_topic_unmuted(stream_id, topic) {
```
Instead of using both `is_topic_unmuted` and `is_topic_muted`, we can use a single function named `topic_has_visibility_policy`.

You can take a look at the [corresponding backend function](https://github.com/zulip/zulip/blob/main/zerver/lib/user_topics.py#L180) for your reference.

This approach saves the work of adding a new function every time a new visibility_policy is added. e.g. `is_topic_followed`.

Ddharmani3 replied:
Sure, I will add a common function in the updated PR.

prakhar1144 reviewed (commented):
Thanks @Ddharmani3 for the updates.

I have left some comments.

Also, i think you can maybe update the commit messages keeping the following point from [zulip docs](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#the-purpose-of-the-commit-description) in consideration:
> Don’t include details that are obvious from looking at the diff for the commit, such as lists of the names of the files or functions that were changed, or the fact that you updated the tests.

Ddharmani3 commented on web/src/user_topics.js:
```diff
@@ -19,45 +19,58 @@ export const visibility_policy = {
-        muted_topics.set(stream_id, sub_dict);
+export function update_user_topics(stream_id, topic, visibility_policy, date_updated) {
+    let sub_dict = all_user_topics.get(stream_id);
+    if (visibility_policy === visibility_policy.INHERIT) {
```
How about `new_visibility_policy` which I have also used in `set_user_topic_visibility_policy` function?

timabbott replied:
I think this name might be better -- the bug is probably that the global map should be called something like `all_visibility_policies`. Let's do a commit to rename that global at the start of the series.

timabbott commented on web/src/user_topics.js:
```diff
@@ -19,45 +19,58 @@ export const visibility_policy = {
-    if (sub_dict) {
-        sub_dict.delete(topic);
+export function get_topic_visibility_policy(stream_id, topic) {
+    if (stream_id === undefined) {
```
It might be better to just not call this function with parameters that aren't valid strings?

timabbott commented on web/src/message_list_data.js:
```diff
@@ -223,7 +223,11 @@ export class MessageListData {
+                    message.stream_id,
+                    message.topic,
+                    user_topics.visibility_policy.MUTED,
+                ) || message.mentioned
```
Reading this commit, I think we actually want to keep a `user_topics.is_topic_muted` function that is a thin wrapper around `user_topics.topic_has_visibility_policy`; each of these blocks is way harder to read than the previous code.

timabbott commented on web/src/popover_menus.js:
```diff
@@ -307,13 +307,35 @@ export function initialize() {
+                    user_topics.visibility_policy.UNMUTED,
+                )
+                    ? user_topics.visibility_policy.INHERIT
+                    : user_topics.visibility_policy.UNMUTED;
```
Long ternary clauses like this are really hard to read. Either use a variable for `is_unmuted`, an if/else block, or both, but don't do this :)

timabbott commented on web/src/popover_menus.js:
```diff
@@ -307,13 +307,35 @@ export function initialize() {
 
-            $popper.one("click", ".sidebar-popover-unmute-topic", () => {
-                muted_topics_ui.unmute_topic(stream_id, topic_name);
+            $popper.one("click", ".sidebar-popover-toggle-mute", () => {
```
I think it'd be better to use a different CSS class name and click handler for the "unmute" and "mute" cases; squashing the toggling behavior together feels hard to read.

prakhar1144 commented on web/src/user_topics.js:
```diff
@@ -58,6 +63,54 @@ export function get_muted_topics() {
+                feedback_widget.dismiss();
+                return;
+            }
+            if (!from_hotkey) {
```
I think the order should be:
```js
if (!from_hotkey) {
      return;
 }

if (visibility_policy === all_visibility_policies.INHERIT) {
     feedback_widget.dismiss();
     return;
}
 ```
 Because, if the request is not via hotkey, there is no point of `feedback.dismiss()` -- right?

Ddharmani3 replied:
Sure, I will wait to address this along with the feedback from @timabbott.

timabbott replied:
I think the code is actually better as written; the `feedback_widget` dismissal is to avoid continuing to show an inaccurate `feedback_widget` state where it's saying you just muted a topic, which you've in fact already changed to not be muted.

And there it doesn't matter whether you used the keyboard on a mouse button to unmute the topic; just whether the feedback widget is visible.

prakhar1144 reviewed (commented):
Suggested a minor tweak, otherwise lgtm.

(ICYMI: My [previous comment](https://github.com/zulip/zulip/pull/24989#pullrequestreview-1377310123) regarding commit messages.)

alya commented:
@Ddharmani3 As per the PR template, please remember to always post still screenshots of any UI changes in the description, not just a video.

alya commented:
Actually, I'm realizing this PR probably doesn't require testing/review from me, as this just adds an option rather than implementing the desired UI for now.

Ddharmani3 commented:
> Actually, I'm realizing this PR probably doesn't require testing/review from me, as this just adds an option rather than implementing the desired UI for now.

Yes, this PR does not implement the UI itself, it lays the groundwork for the desired feature. Once this PR is merged, the work will begin on #24243 to implement the desired UI for this feature.

timabbott commented:
I changed the strings to make (1) the strings outside the development environment special case be unchanged and (2) the special development environment strings be clearer about how they're different from the main ones. We can proceed with building the actual UI on top of this.

With those changes, I've marked this to merge once CI passes, thanks for all the work on this @Ddharmani3 and @prakhar1144!
